manifests/osfamily/windows.pp

```diff
@@ -0,0 +1,37 @@
+class puppet_agent::osfamily::windows {
+
+  $_arch = $::kernelmajversion ?{
+    /^5\.\d+/ => 'x86', # x64 is never allowed on windows 2003
+    default   => $::puppet_agent::arch,
+  }
+
+  $_source = $::puppet_agent::source ? {
+    undef   => "https://downloads.puppetlabs.com/windows/puppet-agent-${_arch}-latest.msi",
+    default => $::puppet_agent::source,
+  }
+
+  $_msi_location = $_source ? {
+    /^puppet:/ => "${tmpdir}\\puppet-agent.msi",
+    default    => $_source,
+  }
+
+  if $_source =~ /^puppet:/ {
+    file{ $_msi_location:
+      source => $_source,
+      before => File["${tmpdir}\\install_puppet.bat"],
+    }
+  }
+
+  $_timestamp = strftime('%Y_%m_%d-%H_%M')
+  $_logfile = "${tmpdir}\\puppet-${_timestamp}-installer.log"
+  debug ("Puppet upgrade log file at ${_logfile}")
+  debug ("Installing puppet from ${_msi_location}")
+  file { "${tmpdir}\\install_puppet.bat":
+    ensure  => file,
+    content => template('puppet_agent/install_puppet.bat.erb')
+  }->
+  exec { 'install_puppet.bat':
+    command   => "${::system32}\\cmd.exe /c start /b ${::system32}\\cmd.exe /c ${tmpdir}\\install_puppet.bat",
+    path      => $::path,
```

Comment:
Guess that makes sense if you want to not use path at all, you would have to set `path => '',` which may not be extremely intuitive. But now that I think about it, that's how things naturally work, so when you do normal tasks you have PATH, it should as well. 

I wonder if this leans more towards deterministic settings.